libs/datamodel/core/src/transform/ast_to_dml/validate/referential_actions.rs

```diff
@@ -1,63 +1,195 @@
-use std::{collections::HashSet, fmt, rc::Rc};
-
-use dml::{datamodel::Datamodel, field::RelationField, model::Model, traits::WithName};
+mod visited_relation;
 
 use crate::{
     ast::Span,
     diagnostics::{DatamodelError, Diagnostics},
     transform::ast_to_dml::validate::RELATION_ATTRIBUTE_NAME,
 };
+use dml::{datamodel::Datamodel, field::RelationField, model::Model, traits::WithName};
+use itertools::Itertools;
+use std::{
+    collections::{BTreeSet, HashMap, HashSet},
+    rc::Rc,
+};
+use visited_relation::VisitedRelation;
 
-/// A linked list structure for visited relation paths.
-#[derive(Debug)]
-struct VisitedRelation<'a> {
-    previous: Option<Rc<VisitedRelation<'a>>>,
-    model_name: &'a str,
-    field_name: &'a str,
-}
+/// Find relations from the given model that end up into the same model,
+/// triggering cascading referential actions.
+pub(crate) fn detect_multiple_cascading_paths(
+    datamodel: &Datamodel,
+    parent_model: &Model,
+    parent_field: &RelationField,
+    span: Span,
+    errors: &mut Diagnostics,
+) {
+    let on_update = parent_field
+        .relation_info
+        .on_update
+        .unwrap_or_else(|| parent_field.default_on_update_action());
+
+    let on_delete = parent_field
+        .relation_info
+        .on_delete
+        .unwrap_or_else(|| parent_field.default_on_delete_action());
+
+    if !on_update.triggers_modification() && !on_delete.triggers_modification() {
+        return;
+    }
+
+    // We don't want to cycle forever.
+    let mut visited = HashSet::new();
+
+    // A graph of relations to traverse.
+    let mut next_relations = Vec::new();
+
+    // Gather all paths from this model to any other model, skipping
+    // cyclical relations
+    let mut paths = Vec::new();
+
+    // Add all relations from current model to the graph vector. At this point
+    // we only care about multiple paths from this model to any other model. We
+    // handle paths that cross, but are not started from here, when calling the
+    // function from corresponding models.
```

Comment:
Is this current? I remember the tests had a case where the chain started before the model actually causing the multiple paths (from multiple relation fields).